### PR TITLE
perf(sync): Fix get ancestor performance issue

### PR DESCRIPTION
**Pull request by TheWaWaR**
- [x] Use skip list to search ancestor of a header
- [x] Adjust block fetch interval IBD/not-IBD
- [x] Add test for skip list

**NOTE**: The [implementation](https://github.com/bitcoin/bitcoin/blob/5d2ccf0ce9ca1571c650a69319fb9c1e0b626ecb/src/chain.cpp#L70-L120) and [test](https://github.com/bitcoin/bitcoin/blob/5d2ccf0ce9ca1571c650a69319fb9c1e0b626ecb/src/test/skiplist_tests.cpp#L13-L44) is almost copy from bitcoin.

### CODE REVIEW

File: sync/src/synchronizer/block_fetcher.rs [hunk 1]
```diff
@@ -168,13 +168,17 @@ where
             let mut inflight = self.synchronizer.peers.blocks_inflight.write();
             let count = MAX_BLOCKS_IN_TRANSIT_PER_PEER
                 .saturating_sub(inflight.peer_inflight_count(&self.peer));
+            let max_height_header = self
```
Comment (doitian) [inline sync/src/synchronizer/block_fetcher.rs]: Why this extra step is required?

File: sync/src/types.rs [hunk 2]
```diff
@@ -507,6 +514,69 @@ impl HeaderView {
     pub fn into_inner(self) -> Header {
         self.inner
     }
+
+    pub fn build_skip<F>(&mut self, mut get_header_view: F)
+    where
+        F: FnMut(&H256) -> Option<HeaderView>,
```
Comment (doitian) [inline sync/src/types.rs]: Why need `FnMut`?

File: sync/src/types.rs [hunk 3]
```diff
@@ -507,6 +514,69 @@ impl HeaderView {
...
+        }
+        let mut number_walk = current.number();
+        while number_walk > number {
+            let number_skip = get_skip_height(number_walk);
+            let number_skip_prev = get_skip_height(number_walk - 1);
+            match current.skip_hash {
+                Some(ref hash)
+                    if number_skip == number
+                        || (number_skip > number
+                            && !(number_skip_prev + 2 < number_skip
```
Comment (doitian) [inline sync/src/types.rs]: parent is better if `number_skip_prev < number_skip`, why need `+2` here?

File: sync/src/types.rs (hunk 2 above)
Comment (TheWaWaR) [reply sync/src/types.rs]: Could be removed.

Comment (TheWaWaR) [reply sync/src/types.rs]: Could be just `Fn`

File: sync/src/synchronizer/block_fetcher.rs (hunk 1 above)
Comment (TheWaWaR) [reply sync/src/synchronizer/block_fetcher.rs]: A minor optimization. typical `get_ancestor` with 30W ancestors need about `60` steps, with this optimization will reduce to less than `(max_height - fixed_last_common_header.number())`

File: sync/src/types.rs (hunk 2 above)
Comment (TheWaWaR) [reply sync/src/types.rs]: If we want test this method (measure steps for example) in the future, we need `FnMut`

File: sync/src/types.rs (hunk 3 above)
Comment (TheWaWaR) [reply sync/src/types.rs]: `number_skip_prev` and `number_skip` are in different parity(even+odd or odd+even), when `number_skip_prev < number_skip` their distance can be 1, 3, 5...

I guess by current `get_skip_height`, neighbors' parents also neighbors is impossible, so their distance `>=3`

Comment (TheWaWaR) [reply sync/src/types.rs]: `number_skip_prev` and `number_skip` are in different parity (odd+even or even+odd), so their distance can be 1,3,5...

I guess by current `get_skip_height`, neighbor's parents also neighbor is impossible, so their distance `>=3`

Comment (TheWaWaR) [reply sync/src/types.rs]: Need a prove here?

**Comment by doitian** [comment]
It's an important performance improvement, could you backport this PR to rc/v0.14.0? I think we need this PR in the next release.

**Comment by TheWaWaR** [comment]
> It's an important performance improvement, could you backport this PR to rc/v0.14.0? I think we need this PR in the next release.

OK

**Pull request merged**